CMSgov/mdct-mcr
# Add quality measures functionality (Nick's version™)

## Description

There were some new design changes from R&D to implement and because Nick was out Friday I took 'em on. Due to the number of comments in the old PR I thought I'd close it and open this new one so the PR could start with a fresh slate.

This PR adds quality measures functionality for the ModalDrawerReportPage and all associated components.

Changes: 

  - Adds Quality Measure-specific designs and functionality for EntityCard, EntityCardTopSection, and EntityCardBottomSection.
  - Adds functionality for creating fields that are repeated. This is because the Quality Measure drawer form has a field which has to be answered once for each plan in the report. Because we can't know at the outset how many plans there are, we have to dynamically render the fields based on the number of plans. This required building a utility and some helper logic.
  - Adds minor new functionality to drawers to handle for forms without fields (in the event that there is only one field type and it is repeating and the entity for which it is to be repeated has no entities in it, e.g. if a form consists of one field that is to be repeated for each plan in a report, but there are no plans added yet).
  - Adds minor new functionality for drawer styling (eyebrow header, new messaging types).
  - Adds tons of new mocked objects for Quality Measures entities and formatted entity data and of course a ton of tests, which account for the majority (60%) of the lines of code.

Closes the following tickets: [MDCT-1965](https://qmacbis.atlassian.net/browse/MDCT-1965), [MDCT-1975](https://qmacbis.atlassian.net/browse/MDCT-1975), [MDCT-1976](https://qmacbis.atlassian.net/browse/MDCT-1976), [MDCT-1977](https://qmacbis.atlassian.net/browse/MDCT-1977), [MDCT-1978](https://qmacbis.atlassian.net/browse/MDCT-1978).

### How to test

1. Test measure creation and hydration (card state: not started)
    - Create a program
    - Do not create any plans. Go straight to Section D: Quality Measures
    - Create a quality measure; verify card creation
    - Edit quality measure; verify hydration of modal and update of card data
    - Verify that card is in incomplete state and shows no highlighted measure sections in the card bottom section
2. Test measure entry and hydration (card state: complete)
    - Open measure drawer to enter data; verify that button is disabled and no form fields appear (because no plans have been created)
    - Nav to Section A: Add plans. Add a plan. Save.
    - Nav to Section D: Quality Measures. 
    - Open measure drawer to enter data for measure; verify that field exists for plan. Enter data and save.
    - Verify that card is in a complete state and shows highlighted measure section for completed plan. 
    - Re-open drawer and verify hydration of drawer data. Change data, save, and verify update of card data.
3. Test subsequent plan addition (card state: partially complete)
    - Nav to Section A: Add plans. Add another plan. Save.
    - Nav to Section D: Quality Measures. Verify that card is now in a partially complete state with a red warning icon in the top left and highlighted sections appearing in the card bottom section for each plan in the report. The newly added plan should be highlighted in red instead of blue.
    - Open the drawer to enter measure data. Enter data for the new plan and save.
    - Verify that card has returned to completed state and shows both completed plans in card bottom section.

### Changed Dependencies

n/a

## Code author checklist

## Reviewer checklist (two different people)
